### PR TITLE
332 shp export use gdal

### PR DESCRIPTION
**Pull request by td928**
#332 

## Shp Export Recap
We first came across the shp export issue when prod1 is depreciated and prod2 does not have gdal installed. We then switched over to using docker container for both gdb and shp file exporting functionality. We also in the process decided to move away from self-hosted runner entirely due to difficulties with postgres installation. But the [docker image](https://hub.docker.com/r/webmapp/gdal-docker/dockerfile) never quite worked for shp file exporting as we have hoped. So this work is essentially making `SHP_export` use the local gdal to export. 

## issue-comment-runner.yml and config.sh
the true work is actually quite minimal with simply moving the `SHP_export` off the docker container and add installation `gdal-bin` in the [issue-comment-runner.yml](https://github.com/NYCPlanning/db-pluto/blob/332-shp-export-use-gdal/.github/workflows/issue-comment-runner.yml). [The same workflow has been tested](https://github.com/NYCPlanning/db-pluto/actions/runs/2347631516) and produced good results. 

## bonus and help wanted for enhancement
a bonus included in this work is in order to simulate environment of github hosted runner as closely as possible. I set up a devcontainer based on the [facdb devcontainer](https://github.com/NYCPlanning/db-facilities/tree/develop/.devcontainer). This definitely was an overkill since the devcontainer for pluto could be much more lightweight because no geosupport or python requirement. But I still think the basic principle is the same. Open to suggestion on how to slim down the container to more bare boned. But considering the time constraint with PLUTO might make sense to let it go for now.

### CODE REVIEW

**Review by AmandaDoyle — COMMENTED** [review]
See one comment / question.  Not a showstopper.  It should be easy to make the change to 22a and rerun.

File: .devcontainer/docker-compose.yml
Comment (AmandaDoyle) [reply .devcontainer/docker-compose.yml]: @td928  You mentioned this yesterday and it didn't register.  We are using Geosupport in the PLUTO build.  See [here](https://github.com/NYCPlanning/db-pluto/blob/main/pluto_build/python/geocode.py).

File: .devcontainer/devcontainer.json
Comment (SashaWeinstein) [inline .devcontainer/devcontainer.json:36]: do we need all these linters? Think we should all settle on one. Bandit is the linter set on my machine where this one set, and I think it's sufficient

Comment (SashaWeinstein) [inline .devcontainer/devcontainer.json:36]: do we need all these linters? Think we should all settle on one. Bandit is the linter set on my machine where this one set, and I think it's sufficient

File: .devcontainer/docker-compose.yml
Comment (td928) [reply .devcontainer/docker-compose.yml]: I see! But I see it is only being used on pre-building step in the processing for [input_pts](https://github.com/NYCPlanning/db-pluto/blob/main/.github/workflows/input_pts.yml). Is this the correct reading?

Comment (SashaWeinstein) [reply .devcontainer/docker-compose.yml]: We delegate the geosupport to docker containers with a `docker run --rm`, so we don't need it in the devcontainer

File: .github/workflows/issue-comment-runner.yml
Comment (td928) [reply .github/workflows/issue-comment-runner.yml:50]: Yeah now it comes back to me that we discussed about the [setup.sh](https://github.com/NYCPlanning/db-pluto/blob/main/pluto_build/bin/setup.sh) which we thought about eliminating entirely. Is it maybe actually good to keep and expand?

File: pluto_build/03_corrections.sh
Comment (td928) [reply pluto_build/03_corrections.sh:1]: yep this only comes up as a hard stop when run locally for some reason. On the remote it threw an error and marched on. I am guessing github action has some advance way to make assumptions when not given correct command?

File: .github/workflows/test.yml
Comment (td928) [reply .github/workflows/test.yml:41]: good question. It was done this way in our other containers so I followed suite. I read [this post](https://superuser.com/questions/70031/what-is-the-difference-between-recommended-and-suggested-packages-ubuntu) about recommended packages and it seems like reasonable to go with the default. I don't if @SashaWeinstein  or @AmandaDoyle have other insights about this.

File: .devcontainer/docker-compose.yml
Comment (td928) [reply .devcontainer/docker-compose.yml]: > We delegate the geosupport to docker containers with a `docker run --rm`, so we don't need it in the devcontainer

good catch. Didn't see this at first.

File: .github/workflows/issue-comment-runner.yml
Comment (SashaWeinstein) [reply .github/workflows/issue-comment-runner.yml:50]: haha yea I guess I'm going back and forth on this. I think expanding setup.sh is better

File: pluto_build/03_corrections.sh
Comment (SashaWeinstein) [reply pluto_build/03_corrections.sh:1]: guess so. good catch

File: .github/workflows/test.yml
Comment (mbh329) [reply .github/workflows/test.yml:41]: I can dig that ethos @SashaWeinstein. @td928 okay cool that makes sense, as long as it works

File: pluto_build/bin/config.sh
Comment (abrieff) [inline pluto_build/bin/config.sh:45]: This doesn't replicate the exact same removes from the previous functionality - may need to be updated to reflect

File: pluto_build/bin/config.sh
Comment (abrieff) [inline pluto_build/bin/config.sh:34]: Now that this export looks so similar to SHP_export there's probably an opportunity for DRYing this, don't know how worth putting the effort in is now

File: pluto_build/05_export.sh
Comment (abrieff) [inline pluto_build/05_export.sh:57]: should this wait still be here?

File: pluto_build/04_archive.sh
Comment (SashaWeinstein) [reply pluto_build/04_archive.sh:37]: Those comments are on purpose, that's a janky fix to work around the fact that the QAQC process writes directly to the publishing database. I wrote issue #317 to work around this problem but haven't implemented a fix yet

File: pluto_build/bin/config.sh
Comment (SashaWeinstein) [reply pluto_build/bin/config.sh:34]: The bash code we maintain is overall extremely not DRY. This is an issue we inherited from the Baiyue/Molly days, and my bash skills have never been good enough to refactor any of it.

File: pluto_build/04_archive.sh
Comment (SashaWeinstein) [reply pluto_build/04_archive.sh:37]: The commented out code is on purpose. Currently the QAQC process writes directly to the publishing database, even on local runs. I opened issue #317 to fix this issue and bring the QAQC process more in line with our other data products

File: .devcontainer/docker-compose.yml
Comment (SashaWeinstein) [inline .devcontainer/docker-compose.yml:11]: I'm curious about this line, I thought we didn't use node and didn't need to install it but maybe I'm wrong?

File: pluto_build/04_archive.sh
Comment (td928) [reply pluto_build/04_archive.sh:37]: Not sure what happened here. @SashaWeinstein have any explanations?

File: .devcontainer/docker-compose.yml
Comment (td928) [reply .devcontainer/docker-compose.yml:11]: Nope I should remove this.

File: pluto_build/04_archive.sh
Comment (SashaWeinstein) [reply pluto_build/04_archive.sh:37]: Te can you see the comment I made?

File: pluto_build/05_export.sh
Comment (abrieff) [reply pluto_build/05_export.sh:57]: I think the wait at the beginning of the upload file is probably accomplishing the same thing?

File: pluto_build/04_archive.sh
Comment (td928) [reply pluto_build/04_archive.sh:37]: must be a refresh issue I didnt see it. Now it makes sense. Should those changes be merged along with this PR?

File: pluto_build/05_export.sh
Comment (SashaWeinstein) [reply pluto_build/05_export.sh:57]: I assumed that when the execution got to that `wait` it would check to make sure all background processes were done. So having it below the export step has it check that the export child processes finish. But maybe that's not actually how it works?

File: pluto_build/04_archive.sh
Comment (SashaWeinstein) [reply pluto_build/04_archive.sh:37]: They should not, they should be reverted before we merge

Comment (SashaWeinstein) [reply pluto_build/04_archive.sh:37]: In my mind this PR was still on the "does it work" stage, and then we will have to make more changes for the "cleanup" stage. But maybe we are on to the "cleanup" stage? To be honest I wasn't following quite close enough

Comment (td928) [reply pluto_build/04_archive.sh]: I think all functionalities have been fully tested by me and Andrew and should be ready to merge once approved. Reverting archive back to main.

File: pluto_build/05_export.sh
Comment (abrieff) [reply pluto_build/05_export.sh:57]: Ok, that makes sense

**Review by abrieff — APPROVED** [review]
Looks good to me - my only thought is a better naming convention for the two sets of docker compose files, like docker-compose.development.yml or docker-compose.github.yml or something so their purpose is more self-explanatory/obvious, but not a blocker

File: pluto_build/bin/setup.sh
Comment (SashaWeinstein) [inline pluto_build/bin/setup.sh:15]: Does this line mean that all subsequent bash commands will be run through the ubuntu container? FacDB is more explicit, using `docker-compose exec <commands>` to run each step via the containers. I figured pluto would have similar syntax as we switch to running on a container even on github actions, but I haven't used docker compose enough to know for sure what commands we need to achieve this

**Comment by td928** [comment]
> Looks good to me - my only thought is a better naming convention for the two sets of docker compose files, like docker-compose.development.yml or docker-compose.github.yml or something so their purpose is more self-explanatory/obvious, but not a blocker

This is a good point. But it raises question whether the way I am calling docker compose command in init would still be robust after changing the name. I didn't even know docker compose file naming can be different. Doing some reading on docker compose documentation now.

**Review by SashaWeinstein — COMMENTED** [review]
The changes all make sense to me. What I understand is that we will run pluto via a this docker container that's a combination of the microsoft vsocde image and the webmapp gdal image for both local development and github actions, and this approach will give maximum consistency. What I'm confused about is how exactly this will be done, what I have to go off of is facdb and this code looks different.

File: pluto_build/bin/setup.sh
Comment (SashaWeinstein) [reply pluto_build/bin/setup.sh:15]: Ok cool. My take here is this is an enormously important line, as it changes the OS on which the entire pipeline is run. It's an easy thing for a engineer who less unfamiliar with docker compose (like me) to miss and then get very confused going forward. I think [facdb's implementation](https://github.com/NYCPlanning/db-facilities/blob/86729a33c528bcdcb83ebbbc3c98a16b6d3bc9ef/facdb.sh#L7)
```
docker-compose exec -T facdb facdb init
```
is easier to understand. I'm not saying we should add in those kinds of statements, but maybe there is a way to make our implementation more explicit for the reader?

**Comment by abrieff** [comment]
> The changes all make sense to me. What I understand is that we will run pluto via a this docker container that's a combination of the microsoft vsocde image and the webmapp gdal image for both local development and github actions, and this approach will give maximum consistency. What I'm confused about is how exactly this will be done, what I have to go off of is facdb and this code looks different.

I think this is slightly off - we have two Dockerfiles, one that runs off of the vscode image and the webmap gdal image for local dev, and one that runs off a basic python image and the webmapp gdal image for github. I don't know whether we could just consolidate on one of those options, or whether that would be preferable.

**Comment by SashaWeinstein** [comment]
Ah ok cool that makes sense. And we think that the vscode image we use for development and the python image we use for production builds will be functionally the same? But the vscode image has features we want for development that we don't need to builds

File: pluto_build/bin/setup.sh
Comment (SashaWeinstein) [reply pluto_build/bin/setup.sh:15]: And how do know that it is working? Is there an action run off the 332 branch that successfully runs in a container?

**Comment by abrieff** [comment]
Outstanding work:

1. Fix Archive PGDump step - think this has to do with the proper arguments not being sent to psql to talk to either dockerized postgres or production database
2. Update Issue Comment Runner to use new docker-compose format
3. Remove Upload/Export steps from test.yml that have been added

**Comment by SashaWeinstein** [comment]
Can I be helpful in knocking off any of these items?

**Comment by abrieff** [comment]
Any of them are up for grabs - we could split them up (1 and 2-3)

**Comment by abrieff** [comment]
There's now also what postgres version this should be built off of lol

**Comment by td928** [comment]
> 

Hey @abrieff trying to catch myself up with the status of this work. I looked at the [most recent github action run](https://github.com/NYCPlanning/db-pluto/runs/7240640167?check_suite_focus=true) and based on my understanding about the error message that our version of the postgres is 11 in this case and the publishing is older. I understand the concerns about having not a most up-to-date postgres version and kind of kicking the problem down the line but I am not sure how reverting to the build gdal solution solves this particular issue.

**Comment by abrieff** [comment]
Sorry, the github action isn't going to tell the full story.
The issue is basically this:

Whichever docker stage is put last is ultimately what base image gets used for the build - this means our image for this was being built off of the gdal-docker image, which was ~4 years old with a postgres-client of 9, not the postgres-client of 12 used to export the QAQC sql. This would fail when trying to interact with the sql files exported by the QAQC code. It's confusing because the issue became about the postgres-client on

As far as I could tell, I was unable to upgrade postgresql-client to a more recent version because the OS was too old - but this might be an avenue to avoid addressing this problem.

If we want to have our image built off of python:latest instead of gdal-docker, we need to do one of two things

1. Copy the relevant installation from the gdal-docker image to the python:latest image, which I was iterating on doing but hitting some dependency issues that maybe could be worked through (two different versions of libcrypto is where I got stuck)
2. Go back to just building gdal ourselves on the python:latest image and ignoring the gdal-docker image except for the hints it gives us about building gdal. This latter option is probably better as it would allow us to upgrade to the latest versions of FGDB API and GDAL, but is quite confusing in its own way.

All of that being said, maybe there is an easier solution here that I'm missing, a fresh set of eyes could be helpful/we should talk this through

**Comment by abrieff** [comment]
some of the more recent work iterating on fixing this problem
https://github.com/NYCPlanning/db-pluto/tree/switch-to-python-latest-from-gdal-docker

**Pull request closed without merging**
